### ISSUE TITLE
Epic: Improved support for user-defined shells (e.g. fish / zsh, etc)

### ISSUE THREAD

**Issue by axonasif** [gitpod-io/gitpod#10105]
Improve the support for user-defined shells such as zsh, fish, etc. 

Includes, but not limited to:

* Improve documentation around `SHELL` environment variable
* Update Gitpod compatibility matrix
* Update default workspaces images
* Update any other parts of the product that assume bash (where possible)

**Relates to:** 
* https://github.com/gitpod-io/gitpod/issues/9019
* https://github.com/gitpod-io/gitpod/issues/13558
* https://github.com/gitpod-io/workspace-images/issues/984
* https://github.com/gitpod-io/gitpod/issues/9531

<br>

<details>
<summary>Original issue description</summary>

## Is your feature request related to a problem? Please describe

It would be nice to have Gitpod switch to the user defined(`terminal.integrated.defaultProfile.linux` ) SHELL after processing the task commands in `bash`.

## Describe the behaviour you'd like

- Imagine I have this on my `.gitpod.yml`:
```yaml
tasks:
  - name: combined tasks
    command: echo "this is a command (combined)"
    before: echo "this is a before (combined)"
```

- Upon workspace IDE start, I would see this on my task terminal:
<img width="680" alt="Screenshot 2022-05-18 at 9 19 58 PM" src="https://user-images.githubusercontent.com/39482679/169078609-b1d6aa1b-b9aa-46b4-91cb-3db348adc1f6.png">

# Proposed solution

So, I think we could just put `; exec <user-shell>`(only if the user is not using bash as default) at the end of the script that we're sending to the task terminal, which could look like:
```bash
 HISTFILE=/workspace/.gitpod/cmd-0 history -r; {
echo "this is a before (combined)"
} && {
echo "this is a command (combined)"
}; exec /usr/bin/fish # This here mainly
```

We might also need to consider having an option for `.gitpod.yml` to disable the switching behavior as someone might not want it to switch on some specific use case.

## Describe alternatives you've considered

There are two ways I know of at this moment:
1. Set `SHELL` env var in https://gitpod.io/variables or set in dockerfile to make the defined `SHELL` globally be used across the workspace, @akosyakov shared this trick 👍 
However, this will cause issues since many of the bash builtins are either not present in other shells or behaves differently. On the other hand, we also have non POSIX shells like `fish`.
2. Append some custom logic(hijacking `PROMPT_COMMAND`) into `$HOME/.bashrc` to run `exec <shell>` upon detection of a task terminal and to switch only after the commands have been run. This is done via utilizing the dotfiles feature of Gitpod [^1]
However, this is not a neat or straightforward method of doing this.

## Additional context

Internal discussion on [Slack](https://gitpod.slack.com/archives/C020VCB0U5A/p1651077346532239?thread_ts=1651050048.096659&cid=C020VCB0U5A).

</details>

[^1]: https://github.com/axonasif/dotfiles/blob/5d50fdc8a6fd171ecf3ddeed6443ca45e1f817f3/src/main.sh#L61

**Comment by loujaybee** [comment]
Agree, as I have this issue/challenge with using fish terminal, however `terminal.integrated.defaultProfile.linux` is VS Code specific, so would be good to have the configuration done somewhere that's IDe agnostic 🙏

**Comment by akosyakov** [comment]
Yes, I think `SHELL` emv var is a still way to go. Honestly even if supervisor reads from `terminal.integrated.defaultProfile.linux` I doubt it will resolve issues with `fish`. 

So the focus should be to resolve:
> However, this will cause issues since many of the bash builtins are either not present in other shells or behaves differently. On the other hand, we also have non POSIX shells like fish.

**Comment by axonasif** [comment]
Just noticed after daily driving `SHELL=/usr/bin/zsh` for a while, if you set your shell to something else other than bash with `SHELL` variable, you miss out on all the workspace customization we do (e.g environment variables, `bashrc.d` hooks).

Perhaps we should switch to other shells from bash env.

**Comment by loujaybee** [comment]
Hey folks, I have broadened the scope of this issue to be an epic to contain other sub issues about improved support for user defined shells. I will start to update the description with other related issues. If you have issues with any shell when using Gitpod, please raise an issue with the specific issue you are facing, and we can then link it here to track general progress towards improved user-defined shell support 🙏
